File: Data Structures & Algorithms/three-integer-sum/submission-6.rs
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut sols = Vec::new();
        nums.sort();

        for (i, &num) in nums.iter().enumerate() {
            if i > 0 && num == nums[i-1] {
                continue
            } else {
                let (mut l, mut r) = (i+1, nums.len()-1);
                let target = -num;
                while l <  r {
                    let eq = nums[l] + nums[r];
                    if eq == target {
                        sols.push(vec![nums[l], nums[r], nums[i]]);
                        r -= 1;
                        l += 1;
                        while l<r && nums[l] == nums[l-1] {
                            l+=1;
                        }
                        while l<r && nums[r] == nums[r+1] {
                            r-=1;
                        }
                    } else {
                        if eq > target {
                            r -= 1;
                        } else {
                            l += 1;
                        }
                    }
                }
            }
        }
        return sols
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-6.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        l,r = 0,len(matrix)-1

        while l<=r:
            m = (r+l)//2
            if matrix[m][-1] < target:
                l=m+1
            elif matrix[m][0] > target:
                r=m-1
            else:
                sl,sr = 0,len(matrix[m])-1
                while sl <= sr:
                    sm = (sl+sr)//2
                    if matrix[m][sm] < target:
                        sl = sm+1
                    elif matrix[m][sm] > target:
                        sr = sm-1
                    else:
                        return True
                return False
        return False
            


